File: src/python/clock.py
import time

NANO = 1000000000.0

class Clock:
    def __init__(self):
        self.start_time = 0
        self.stop_time = 0
        self.running = False
        
    
    def reset(self):
        self.start_time = 0
        self.running = False
    
    def start(self):
        self.start_time = time.time_ns()
        self.running = True
    
    def stop(self):
        if self.running:
            self.stop_time = time.time_ns()
            self.running = False
    
    def get_time(self) -> float:
        if self.running:
            elapsed = (time.time_ns() - self.start_time) / NANO
        else:
            elapsed = (self.stop_time - self.start_time) / NANO
        return elapsed

File: src/rust/dpll/src/clock.rs
use std::time::Instant;

const NANO: f64 = 1_000_000_000.0;

pub struct Clock {
    start_time: Option<Instant>,
    stop_time: Option<Instant>,
    running: bool,
}

impl Clock {
    pub fn new() -> Self {
        Clock {
            start_time: None,
            stop_time: None,
            running: false,
        }
    }

    pub fn reset(&mut self) {
        self.start_time = None;
        self.stop_time = None;
        self.running = false;
    }

    pub fn start(&mut self) {
        self.start_time = Some(Instant::now());
        self.running = true;
    }

    pub fn stop(&mut self) {
        if self.running {
            self.stop_time = Some(Instant::now());
            self.running = false;
        }
    }

    pub fn get_time(&self) -> f64 {
        match self.running {
            true => {
                if let Some(start_time) = self.start_time {
                    let elapsed = start_time.elapsed();
                    elapsed.as_nanos() as f64 / NANO
                } else {
                    0.0
                }
            },
            false => {
                if let Some(start_time) = self.start_time {
                    if let Some(stop_time) = self.stop_time {
                        let elapsed = stop_time.duration_since(start_time);
                        elapsed.as_nanos() as f64 / NANO
                    } else {
                        0.0
                    }
                } else {
                    0.0
                }
            }
        }
    }
}